add getConfig and fix return value of setConfig

atari2600tim opened the pull request:
<!-- Keep the title short & concise so anyone non-technical can understand it,
     the title appears in PTB changelogs -->
#### Brief overview of PR changes/additions
Check for configuration options similar to those that are set by setConfig.

Also made setConfig return the results instead of dropping them.

#### Motivation for adding to Mudlet
There is already setConfig so that I can disable "Fix unnecessary linebreaks on GA servers", but I couldn't check the current status to suggest changing it, notify that script is changing it, or to refrain from messaging them if it's already as suggested.  Also `getConfig("enabledGMCP")` resolves #1729 where they want to see if Enable GMCP is checked.

#### Other info (issues closed, discussion etc)
I did all the same options except for showMapInfo and hideMapInfo because they work different.

atari2600tim committed:
initial

atari2600tim committed:
Merge branch 'Mudlet:development' into tim-getConfig

atari2600tim committed:
missed registering the function, plus formatting

atari2600tim committed:
Merge branch 'tim-getConfig' of https://github.com/atari2600tim/Mudlet into tim-getConfig

atari2600tim committed:
spacing

atari2600tim committed:
Merge branch 'Mudlet:development' into tim-getConfig

atari2600tim committed:
Update TLuaInterpreter.h

token edit after ubuntu builder is supposed to be back

atari2600tim committed:
Update TLuaInterpreter.h

Kebap commented on src/TLuaInterpreter.cpp:
```diff
@@ -17870,3 +17871,128 @@ int TLuaInterpreter::getScroll(lua_State* L)
+    const bool currentHost = (mudlet::self()->mpCurrentActiveHost == &host);
+    lua_newtable(L);
+    
+    if (host.mpMap && host.mpMap->mpMapper) {
```
This check seems unneccessary to me here, except for the show3dMapView value case.
The other values could be retrieved from host, even if no mpMap or mpMapper exist?

Kebap commented:
Could you kindly add brief comments to this source and `TLuaInterpreter::setConfig` to the effect, if you add an option to any of them, make sure to also add it to the other command, to keep them in sync? Thanks in advance!

atari2600tim committed:
show map settings, comment to keep sync

show map settings even though they may not be settable at the moment
I think all the settings have defaults in host and do not rely on the mapper opening to set them.

atari2600tim commented:
I notice that the wiki regarding setConfig says that it returns `true if successful, or nil+msg if the option doesn't exist`.

The `TLuaInterpreter::setConfig` does that.  Then the code in `Other.lua` takes that result and ignores it and returns nothing.

Now, if it is given a key and value like `setConfig("mapRoomSize", 5)`, that is obvious enough, I should return what the C function returned.  But what if it was given a table?  If you run `setConfig({mapRoomSize = 6, frog = 12})` then should it return true because a change was made to mapRoomSize, or should it return nil+msg because frog is not something that can be set?  Just return nothing in case of tables?  Return true if there was at least one success?  Return nil+msg if there was at least one failure?

atari2600tim committed:
refrain from dropping setConfig result

vadi2 commented:
Good point and this is a pattern that we haven't considered before - what to do in case of batched input. @Mudlet/lua-interface ?

vadi2 commented:
For the bounty we can put this aside since it's an issue with setConfig as well, but we do need to find a good answer to this.

keneanung commented:
I created #6569 for the discussion to not dilute the PR.

atari2600tim commented:
>it's an issue with setConfig as well

It's only an issue with the set side, and at the risk of scope creep I just moved the return up one line to solve it instead of making a whole PR just for that.  The table format of parameter though is still left alone and returning nothing.  In retrospect since I did have a question, a separate issue like 6569 is better..

Get doesn't offer a table format, where you could tell it 2 things to get the results filtered down to just those 2.  Should it? 
 Wouldn't be difficult but is a minor complexity to throw in.

Kebap commented:
> Get doesn't offer a table format, where you could tell it 2 things to get the results filtered down to just those 2. Should it?
> Wouldn't be difficult but is a minor complexity to throw in.

I think it would be a good addition, because your current code would otherwise fetch all results, throw away all but one, then fetch all results again, and throw away all but the other. That does not sound very efficient, no.

Probably it could be worthwhile to also inform the C++-function about which (singular or multiple) result is required, so it doesn't even produce the whole list, which are then mostly ignored by the superimposed Lua-function afterwards automatically.

atari2600tim commented:
I was thinking of how useless this would be:
```
local result = getConfig({"enabledGMCP", "enabledMSSP"})
if result["enabledGMCP"] then echo("you have gmcp on") end
if result["enabledMSSP"] then echo("you have mssp on") end
```
You could just as easily grab same thing from full list.

But at same time, maybe you're not using a code box, maybe you're using the lua alias, in which case having it filtered would help.

Initially I started off with building the list by having the C function be aware of what they are wanting.  I didn't like any of the logic I came up with.  

First I had something like
```
if zero args then {
  put "enableGMCP" on the stack, then the value
  put "enableMSSP" on the stack, then the value
  do what I made the PR do
}
else{
  if arg == "enableGMCP" {
    return it
  } else {
    return error
  }
}
```
That didn't seem right because it had 2 places that were both checking what value to put for that setting... it didn't run both, not slower but just bloated, two paths each did same thing, and whoever updates it needs to update 2 things.
Then I had something like 
```
if zero args then { fullList = true }
else { argName = arg }
start lua table
if (fullList || arg=="enableGMCP") {
  add "enableGMCP"
  add whatever boolean
}
end lua table
if fullList {
  pop that table off and replace it with just the value
}
```
Doing it that way, it ended up only doing one code block, but needlessly created and removed a table in the case where they requested a particular thing

I didn't like any of those above and just came up with the full list building.  But now as I'm writing this I thought of a different option.  Have it ONLY get one thing.  If you want the whole set then have the Other.lua run a loop to build a full list in same way that the setter does.  I'll go ahead and try that out.

atari2600tim committed:
change main function to get one at a time

vadi2 commented on src/mudlet-lua/lua/Other.lua:
```diff
@@ -1170,11 +1170,58 @@ function setConfig(...)
+      "caretShortcut", 
+    }
+    for _,v in ipairs(list) do
+      result[v]=oldgetConfig(v)
```
```suggestion
      result[v] = oldgetConfig(v)
```

vadi2 commented on src/mudlet-lua/lua/Other.lua:
```diff
@@ -1170,11 +1170,58 @@ function setConfig(...)
+    return result
+  end
+
+  if type(args[1])== "table" then
```
```suggestion
  if type(args[1]) == "table" then
```

vadi2 reviewed (approved):
This is going above & beyond. Thanks!

Kebap commented on src/TLuaInterpreter.cpp:
```diff
@@ -17876,132 +17876,159 @@ int TLuaInterpreter::getScroll(lua_State* L)
 
 // Documentation: https://wiki.mudlet.org/w/Manual:Lua_Functions#getConfig
 // Please use same options with same names in setConfig and getConfig and keep them in sync
+// Also list is in Other.lua
```
Now also keep comment in sync with setConfig please 😄

atari2600tim replied:
Now they each have an explanation of the front-end that is placed before them for the multiple variable cases.

atari2600tim committed:
Apply suggestions from code review

Co-authored-by: Vadim Peretokin <vperetokin@hey.com>

atari2600tim committed:
Update TLuaInterpreter.cpp

vadi2 commented:
Thanks @atari2600tim!
